### PR TITLE
Use explicit API

### LINKED ISSUE #609: Use explicit API
## Feature Request

- Make the API as explicit as possible with typehints
- Close all classes
- Prefer `private` visibility

### PR DESCRIPTION
**Pull request by phansys**
## Subject

Use explicit API

I am targeting this branch, because these changes break BC.

Triggered by #609.

## To do
    
- [x] Fix tests.

### CODE REVIEW

File: src/Block/BlockContextManager.php [hunk 1]
```diff
@@ -60,10 +60,7 @@ public function __construct(BlockLoaderInterface $blockLoader, BlockServiceManag
         $this->logger = $logger;
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function addSettingsByType($type, array $settings, $replace = false): void
+    public function addSettingsByType(string $type, array $settings, $replace = false): void
```
Comment (core23) [inline src/Block/BlockContextManager.php]:
```suggestion
    public function addSettingsByType(string $type, array $settings, bool $replace = false): void
```

File: src/Block/BlockContextManager.php [hunk 2]
```diff
@@ -73,10 +70,7 @@ public function addSettingsByType($type, array $settings, $replace = false): voi
         }
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function addSettingsByClass($class, array $settings, $replace = false): void
+    public function addSettingsByClass(string $class, array $settings, $replace = false): void
```
Comment (core23) [inline src/Block/BlockContextManager.php]:
```suggestion
    public function addSettingsByClass(string $class, array $settings, bool $replace = false): void
```

File: src/Block/BlockLoaderChain.php [hunk 3]
```diff
@@ -62,10 +58,7 @@ public function load($block)
         throw new BlockNotFoundException();
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function support($name)
+    public function support($name): bool
```
Comment (core23) [inline src/Block/BlockLoaderChain.php]:
```suggestion
    public function support(string $name): bool
```

File: src/Block/BlockLoaderChain.php [hunk 4]
```diff
@@ -48,10 +47,7 @@ public function exists($type)
         return false;
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function load($block)
+    public function load($block): BlockInterface
```
Comment (core23) [inline src/Block/BlockLoaderChain.php]:
```suggestion
    public function load(string $block): BlockInterface
```

**Comment by phansys** [comment]
Should we still allow to hold service ids instead of service instances at `BlockServiceManager`?
If yes, I must revert some type declarations.

https://github.com/sonata-project/SonataBlockBundle/blob/d0a67fcd6873e234467cd58b35aa3feb5b54a6ff/src/Block/BlockServiceManager.php#L201-L207

File: src/Cache/HttpCacheHandlerInterface.php [hunk 5]
```diff
@@ -30,8 +30,6 @@ public function alterResponse(Response $response);
      * This function can update a state to store the final ttl used for the page
      * The response object point to the Response generated by a block service and
      * not the final response rendered to the client.
-     *
-     * @param BlockContextInterface $blockContext
      */
     public function updateMetadata(Response $response, BlockContextInterface $blockContext = null);
```
Comment (core23) [inline src/Cache/HttpCacheHandlerInterface.php]:
```suggestion
    public function updateMetadata(Response $response, BlockContextInterface $blockContext = null) : void;
```

File: src/Exception/Filter/FilterInterface.php [hunk 6]
```diff
@@ -27,10 +27,10 @@ interface FilterInterface
...
      * Returns whether or not this filter handles this exception for given block.
      *
-     * @param \Exception     $exception Exception to manage
+     * @param \Throwable     $exception Exception to manage
      * @param BlockInterface $block     Block that provoked the exception
      *
      * @return bool
      */
-    public function handle(\Exception $exception, BlockInterface $block);
+    public function handle(\Throwable $exception, BlockInterface $block);
```
Comment (core23) [inline src/Exception/Filter/FilterInterface.php]:
```suggestion
    public function handle(\Throwable $exception, BlockInterface $block): bool;
```

File: src/Exception/Filter/DebugOnlyFilter.php [hunk 7]
```diff
@@ -27,21 +27,18 @@ final class DebugOnlyFilter implements FilterInterface
...
+    public function __construct(bool $debug)
     {
         $this->debug = $debug;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function handle(\Exception $exception, BlockInterface $block)
+    public function handle(\Throwable $exception, BlockInterface $block)
```
Comment (core23) [inline src/Exception/Filter/DebugOnlyFilter.php]:
```suggestion
    public function handle(\Throwable $exception, BlockInterface $block): bool
```

File: src/Exception/Filter/FilterInterface.php (hunk 6 above)
Comment (core23) [reply src/Exception/Filter/FilterInterface.php]: Same for the other filter...

File: src/Block/BlockLoaderChain.php [hunk 8]
```diff
@@ -48,11 +47,12 @@ public function exists($type)
...
     {
+        if (!\is_string($block) && !\is_array($block)) {
+            throw new \TypeError(sprintf('Argument 1 passed to %s must be of type string or array, %s given', __METHOD__, \is_object($block) ? 'object of type '.\get_class($block) : \gettype($block)));
```
Comment (greg0ire) [inline src/Block/BlockLoaderChain.php]:
```suggestion
            throw new \TypeError(sprintf(
                'Argument 1 passed to %s must be of type string or array, %s given',
                __METHOD__,
                \is_object($block) ? 'object of type '.\get_class($block) : \gettype($block)
            ));
```

File: src/Block/BlockLoaderChain.php [hunk 9]
```diff
@@ -62,11 +62,12 @@ public function load($block)
...
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function support($name)
+    public function support($name): bool
     {
+        if (!\is_string($name) && !\is_array($name)) {
+            throw new \TypeError(sprintf('Argument 1 passed to %s must be of type string or array, %s given', __METHOD__, \is_object($name) ? 'object of type '.\get_class($name) : \gettype($name)));
```
Comment (greg0ire) [inline src/Block/BlockLoaderChain.php]:
```suggestion
            throw new \TypeError(sprintf(
                'Argument 1 passed to %s must be of type string or array, %s given',
                __METHOD__, \is_object($name) ? 'object of type '.\get_class($name) : \gettype($name)
            ));
```

File: src/Block/BlockRenderer.php [hunk 10]
```diff
@@ -63,18 +63,15 @@ final class BlockRenderer implements BlockRendererInterface
      * @param LoggerInterface              $logger                   Logger class
      * @param bool                         $debug                    Whether in debug mode or not
      */
-    public function __construct(BlockServiceManagerInterface $blockServiceManager, StrategyManagerInterface $exceptionStrategyManager, LoggerInterface $logger = null, $debug = false)
+    public function __construct(BlockServiceManagerInterface $blockServiceManager, StrategyManagerInterface $exceptionStrategyManager, LoggerInterface $logger = null, bool $debug = false)
```
Comment (greg0ire) [inline src/Block/BlockRenderer.php]:
```suggestion
    public function __construct(
        BlockServiceManagerInterface $blockServiceManager,
        StrategyManagerInterface $exceptionStrategyManager,
        LoggerInterface $logger = null,
        bool $debug = false
    )
```

File: src/Block/BlockRenderer.php [hunk 11]
```diff
@@ -134,11 +120,8 @@ private function createResponse(BlockContextInterface $blockContext, Response $r
 
     /**
      * This method is responsible to cascade ttl to the parent block.
-     *
-     *
-     * @return Response
      */
-    private function addMetaInformation(Response $response, BlockContextInterface $blockContext, BlockServiceInterface $service)
+    private function addMetaInformation(Response $response, BlockContextInterface $blockContext, BlockServiceInterface $service): Response
```
Comment (greg0ire) [inline src/Block/BlockRenderer.php]:
```suggestion
    private function addMetaInformation(
        Response $response,
        BlockContextInterface $blockContext,
        BlockServiceInterface $service
    ): Response
```

File: src/Block/Loader/ServiceLoader.php [hunk 12]
```diff
@@ -35,19 +36,18 @@ public function __construct(array $types)
...
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function load($configuration)
+    public function load($configuration): BlockInterface
     {
+        if (!\is_string($configuration) && !\is_array($configuration)) {
+            throw new \TypeError(sprintf('Argument 1 passed to %s must be of type string or array, %s given', __METHOD__, \is_object($configuration) ? 'object of type '.\get_class($configuration) : \gettype($configuration)));
```
Comment (greg0ire) [inline src/Block/Loader/ServiceLoader.php]:
```suggestion
            throw new \TypeError(sprintf(
                'Argument 1 passed to %s must be of type string or array, %s given',
                __METHOD__,
                \is_object($configuration) ? 'object of type '.\get_class($configuration) : \gettype($configuration)
            ));
```

File: src/Block/Loader/ServiceLoader.php [hunk 13]
```diff
@@ -66,11 +66,12 @@ public function load($configuration)
...
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function support($configuration)
+    public function support($configuration): bool
     {
+        if (!\is_string($configuration) && !\is_array($configuration)) {
+            throw new \TypeError(sprintf('Argument 1 passed to %s must be of type string or array, %s given', __METHOD__, \is_object($configuration) ? 'object of type '.\get_class($configuration) : \gettype($configuration)));
```
Comment (greg0ire) [inline src/Block/Loader/ServiceLoader.php]:
```suggestion
            throw new \TypeError(sprintf(
                'Argument 1 passed to %s must be of type string or array, %s given',
                __METHOD__,
                \is_object($configuration) ? 'object of type '.\get_class($configuration) : \gettype($configuration)
            ));
```

File: src/Block/Service/MenuBlockService.php [hunk 14]
```diff
@@ -48,31 +48,15 @@ final class MenuBlockService extends AbstractAdminBlockService
      */
     protected $menuRegistry;
 
-    /**
-     * @param MenuRegistryInterface|null $menuRegistry
-     */
-    public function __construct(string $name, Environment $twig, MenuProviderInterface $menuProvider, $menuRegistry = null)
+    public function __construct(string $name, Environment $twig, MenuProviderInterface $menuProvider, MenuRegistryInterface $menuRegistry = null)
```
Comment (greg0ire) [inline src/Block/Service/MenuBlockService.php]:
```suggestion
    public function __construct(
        string $name,
        Environment $twig,
        MenuProviderInterface $menuProvider,
        ?MenuRegistryInterface $menuRegistry = null
    )
```

File: src/Meta/Metadata.php [hunk 15]
```diff
@@ -46,13 +46,9 @@ final class Metadata implements MetadataInterface
...
 
     /**
-     * @param string               $title
-     * @param string               $description
-     * @param string|null          $image
-     * @param string|null          $domain
      * @param array<string, mixed> $options
      */
-    public function __construct($title, $description = null, $image = null, $domain = null, array $options = [])
+    public function __construct(string $title, string $description = null, string $image = null, string $domain = null, array $options = [])
```
Comment (greg0ire) [inline src/Meta/Metadata.php]:
```suggestion
    public function __construct(
        string $title,
        string $description = null,
        string $image = null,
        string $domain = null,
        array $options = []
    )
```

File: src/Meta/Metadata.php [hunk 16]
```diff
@@ -46,47 +46,48 @@ final class Metadata implements MetadataInterface
...
-     * @param string               $description
-     * @param string|null          $image
-     * @param string|null          $domain
      * @param array<string, mixed> $options
      */
-    public function __construct($title, $description = null, $image = null, $domain = null, array $options = [])
-    {
+    public function __construct(
+        string $title,
+        string $description = null,
```
Comment (core23) [inline src/Meta/Metadata.php]:
```suggestion
        ?string $description = null,
```

Also for the other parameters please

File: src/Exception/Strategy/StrategyManager.php [hunk 17]
```diff
@@ -112,10 +108,7 @@ public function setDefaultRenderer($name): void
         $this->defaultRenderer = $name;
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function handleException(\Exception $exception, BlockInterface $block, Response $response = null)
+    public function handleException(\Throwable $exception, BlockInterface $block, Response $response = null): Response
```
Comment (core23) [inline src/Exception/Strategy/StrategyManager.php]:
```suggestion
    public function handleException(\Throwable $exception, BlockInterface $block, ?Response $response = null): Response
```

File: src/Block/Service/TextBlockService.php [hunk 18]
```diff
@@ -29,10 +29,7 @@
  */
 final class TextBlockService extends AbstractAdminBlockService
 {
-    /**
-     * {@inheritdoc}
-     */
-    public function execute(BlockContextInterface $blockContext, Response $response = null)
+    public function execute(BlockContextInterface $blockContext, Response $response = null): Response
```
Comment (core23) [inline src/Block/Service/TextBlockService.php]:
```suggestion
    public function execute(BlockContextInterface $blockContext, ?Response $response = null): Response
```

Also for the other blocks...

File: src/Block/Service/MenuBlockService.php [hunk 19]
```diff
@@ -48,31 +48,19 @@ final class MenuBlockService extends AbstractAdminBlockService
...
-    /**
-     * @param MenuRegistryInterface|null $menuRegistry
-     */
-    public function __construct(string $name, Environment $twig, MenuProviderInterface $menuProvider, $menuRegistry = null)
-    {
+    public function __construct(
+        string $name,
+        Environment $twig,
+        MenuProviderInterface $menuProvider,
+        MenuRegistryInterface $menuRegistry = null
```
Comment (core23) [inline src/Block/Service/MenuBlockService.php]:
```suggestion
        ?MenuRegistryInterface $menuRegistry = null
```

File: src/Block/Service/BlockServiceInterface.php [hunk 20]
```diff
@@ -23,27 +23,16 @@
  */
 interface BlockServiceInterface
 {
-    /**
-     * @param Response $response
-     *
-     * @return Response
-     */
-    public function execute(BlockContextInterface $blockContext, Response $response = null);
+    public function execute(BlockContextInterface $blockContext, Response $response = null): Response;
```
Comment (core23) [inline src/Block/Service/BlockServiceInterface.php]:
```suggestion
    public function execute(BlockContextInterface $blockContext, ?Response $response = null): Response;
```

File: src/Block/Service/AbstractBlockService.php [hunk 21]
```diff
@@ -82,13 +77,8 @@ public function renderResponse($view, array $parameters = [], Response $response
...
      * Returns a Response object that cannot be cacheable, this must be used if the Response is related to the user.
      * A good solution to make the page cacheable is to configure the block to be cached with javascript ...
-     *
-     * @param string   $view
-     * @param Response $response
-     *
-     * @return Response
      */
-    public function renderPrivateResponse($view, array $parameters = [], Response $response = null)
+    public function renderPrivateResponse(string $view, array $parameters = [], Response $response = null): Response
```
Comment (core23) [inline src/Block/Service/AbstractBlockService.php]:
```suggestion
    public function renderPrivateResponse(string $view, array $parameters = [], ?Response $response = null): Response
```

File: src/Block/Loader/ServiceLoader.php [hunk 22]
```diff
@@ -35,19 +36,22 @@ public function __construct(array $types)
...
         return \in_array($type, $this->types, true);
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function load($configuration)
+    public function load($configuration): BlockInterface
     {
+        if (!\is_string($configuration) && !\is_array($configuration)) {
```
Comment (core23) [inline src/Block/Loader/ServiceLoader.php]: Duplicate code. Can't you just call the `support` method?

File: src/Block/BlockRenderer.php [hunk 23]
```diff
@@ -113,12 +108,7 @@ public function render(BlockContextInterface $blockContext, Response $response =
         return $response;
     }
 
-    /**
-     * @param Response $response
-     *
-     * @return Response
-     */
-    private function createResponse(BlockContextInterface $blockContext, Response $response = null)
+    private function createResponse(BlockContextInterface $blockContext, Response $response = null): Response
```
Comment (core23) [inline src/Block/BlockRenderer.php]:
```suggestion
    private function createResponse(BlockContextInterface $blockContext, ?Response $response = null): Response
```

File: src/Block/BlockRenderer.php [hunk 24]
```diff
@@ -63,18 +63,19 @@ final class BlockRenderer implements BlockRendererInterface
      * @param LoggerInterface              $logger                   Logger class
      * @param bool                         $debug                    Whether in debug mode or not
      */
-    public function __construct(BlockServiceManagerInterface $blockServiceManager, StrategyManagerInterface $exceptionStrategyManager, LoggerInterface $logger = null, $debug = false)
-    {
+    public function __construct(
+        BlockServiceManagerInterface $blockServiceManager,
+        StrategyManagerInterface $exceptionStrategyManager,
+        LoggerInterface $logger = null,
```
Comment (core23) [inline src/Block/BlockRenderer.php]:
```suggestion
        ?LoggerInterface $logger = null,
```

File: src/Block/Loader/ServiceLoader.php (hunk 22 above)
Comment (phansys) [reply src/Block/Loader/ServiceLoader.php]: I didn't perform that simplification in order to not change the call flow (`load()` and `support()`). If it doesn't worth, I can make the update as suggested.

File: src/Form/Mapper/FormMapper.php [hunk 25]
```diff
@@ -20,7 +20,7 @@
  */
 interface FormMapper
 {
-    public function create(string $name, ?string $type = null, array $options = []): FormBuilderInterface;
+    public function create(string $name, ??string $type = null, array $options = []): FormBuilderInterface;
```
Comment (greg0ire) [inline src/Form/Mapper/FormMapper.php]: :thinking:

File: src/Form/Mapper/FormMapper.php [hunk 26]
```diff
@@ -30,7 +30,7 @@ public function reorder(array $keys): self;
     /**
      * @param FormBuilderInterface|string $name
      */
-    public function add($name, ?string $type = null, array $options = []): self;
+    public function add($name, ??string $type = null, array $options = []): self;
```
Comment (greg0ire) [inline src/Form/Mapper/FormMapper.php]: nope

File: src/Model/BlockInterface.php [hunk 27]
```diff
@@ -127,7 +127,7 @@ public function setSetting(string $name, $value): void;
      *
      * @return mixed
      */
-    public function getSetting(string $name, $default = null);
+    public function getSetting(string ?$name, $default = null);
```
Comment (greg0ire) [inline src/Model/BlockInterface.php]: wut

File: src/Model/BlockInterface.php [hunk 28]
```diff
@@ -85,7 +85,7 @@ public function getCreatedAt(): ?\DateTime;
     /**
      * Set the last update date and time.
      */
-    public function setUpdatedAt(\DateTime $updatedAt = null): void;
+    public function ?setUpdatedAt(\DateTime $updatedAt = null): void;
```
Comment (greg0ire) [inline src/Model/BlockInterface.php]: lol no

File: src/Model/BlockInterface.php [hunk 29]
```diff
@@ -75,7 +75,7 @@ public function getPosition(): ?int;
     /**
      * Sets the creation date and time.
      */
-    public function setCreatedAt(\DateTime $createdAt = null): void;
+    public function ?setCreatedAt(\DateTime $createdAt = null): void;
```
Comment (greg0ire) [inline src/Model/BlockInterface.php]: Rogue script?

File: src/Templating/Helper/BlockHelper.php [hunk 30]
```diff
@@ -84,7 +84,7 @@ class BlockHelper
 
     public function __construct(BlockServiceManagerInterface $blockServiceManager, array $cacheBlocks, BlockRendererInterface $blockRenderer,
                                 BlockContextManagerInterface $blockContextManager, EventDispatcherInterface $eventDispatcher,
-                                CacheManagerInterface $cacheManager = null, HttpCacheHandlerInterface $cacheHandler = null, Stopwatch $stopwatch = null)
+                                ?CacheManagerInterface $cacheManager = null, ?HttpCacheHandlerInterface $cacheHandler = null, ?Stopwatch $stopwatch = null)
```
Comment (greg0ire) [inline src/Templating/Helper/BlockHelper.php]: Please use a multiline signature that is not in a pythonish style

File: src/Form/Mapper/FormMapper.php (hunk 25 above)
Comment (phansys) [reply src/Form/Mapper/FormMapper.php]: Bad regex replace :disappointed:
It should be fixed now.

**Comment by greg0ire** [comment]
Thanks @phansys !

**Pull request merged**